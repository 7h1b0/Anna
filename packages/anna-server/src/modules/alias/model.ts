import Ajv from 'ajv';
import { v4 as uuidv4 } from 'uuid';
import knex from '../../knexClient';
import aliasSchema from './schema';
import { omit } from '../../utils/utils';

export const TABLE = 'alias';
export const COLUMNS = [
  'aliasId',
  'sceneId',
  'name',
  'description',
  'enabled',
  'startTime',
  'endTime',
  'createdAt',
  'updatedAt',
  'createdBy',
];

type Alias = {
  aliasId: string;
  sceneId: string;
  name: string;
  description?: string;
  startTime?: number;
  endTime?: number;
  enabled: boolean;
  createdAt: number;
  updatedAt: number;
  createdBy: string;
};

export function validate(data: Record<string, unknown>) {
  const ajv = new Ajv();
  return ajv.validate(aliasSchema, data);
}

export async function findAll(): Promise<Alias[]> {
  return knex(TABLE).select(COLUMNS);
}

export async function findById(aliasId: string): Promise<Alias> {
  return knex(TABLE).first(COLUMNS).where('aliasId', aliasId);
}

export async function findByName(name: string): Promise<Alias[]> {
  return knex(TABLE).select(COLUMNS).where('name', name);
}

export async function save({
  sceneId,
  name,
  description,
  createdBy,
  enabled = true,
  startTime,
  endTime,
}: {
  sceneId: string;
  name: string;
  description: string;
  createdBy: string;
  enabled?: boolean;
  startTime?: number;
  endTime?: number;
}): Promise<string> {
  const aliasId = uuidv4();
  await knex(TABLE).insert({
    aliasId,
    sceneId,
    description,
    name,
    enabled,
    createdBy,
    startTime,
    endTime,
  });
  return aliasId;
}

export function remove(aliasId: string) {
  return knex(TABLE).where('aliasId', aliasId).del();
}

export function findByIdAndUpdate(aliasId: string, payload: Partial<Alias>) {
  const safePayload = omit(
    payload,
    'aliasId',
    'createdAt',
    'updatedAt',
    'createdBy',
  );
  return knex(TABLE).update(safePayload).where('aliasId', aliasId);
}

export function isInteger(value: unknown | number): value is number {
  return Number.isInteger(value);
}

/**
 * Given aliases, returns one alias respecting the constraints
 */
export function resolveActiveAlias(aliases: Alias[]): Alias | undefined {
  const time = new Date().getHours();
  const activeAlias = aliases
    .filter((alias) => alias.enabled)
    .filter((alias) => {
      if (isInteger(alias.startTime) && isInteger(alias.endTime)) {
        if (alias.startTime < alias.endTime) {
          return time >= alias.startTime && time < alias.endTime;
        }
        return time >= alias.startTime || time < alias.endTime;
      }
      return true;
    });

  return activeAlias.pop();
}
