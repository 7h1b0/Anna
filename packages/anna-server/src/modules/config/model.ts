import knex from '../../knexClient';

export const TABLE = 'config';

export type Config = {
  isAway: boolean;
};

export async function fetchIsAway(): Promise<{ isAway: true }> {
  return knex(TABLE).first('isAway');
}

export function updateIsAway(isAway: boolean) {
  return knex(TABLE).update({ isAway });
}
