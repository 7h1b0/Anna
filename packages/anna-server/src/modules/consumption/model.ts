import knex from '../../knexClient';

export const TABLE = 'consumption';
export const COLUMNS = ['date', 'value'];

export type Consumption = {
  date: number;
  value: number;
};

export async function findLastEntries(limit: number): Promise<Consumption[]> {
  return knex(TABLE)
    .select(COLUMNS)
    .whereNotNull('value')
    .orderBy('date', 'desc')
    .limit(limit);
}

export async function findLastEntry(): Promise<Consumption> {
  return knex(TABLE)
    .first(COLUMNS)
    .whereNotNull('value')
    .orderBy('date', 'desc')
    .limit(1);
}

export async function save(consumptions: { date: Date; value: number }[]) {
  return knex.transaction((trx) => {
    return Promise.all(
      consumptions.map((consumption) => trx.insert(consumption).into(TABLE)),
    );
  });
}
