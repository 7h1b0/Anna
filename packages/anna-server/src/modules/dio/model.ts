import Ajv from 'ajv';
import knex from '../../knexClient';
import dioSchema from './schema';

export const TABLE = 'dios';

export type Dio = {
  dioId: number;
  roomId: string;
  name: string;
};

export function validate(data: Record<string, unknown>) {
  const ajv = new Ajv();
  return ajv.validate(dioSchema, data);
}

export async function findAll(): Promise<Dio[]> {
  return knex(TABLE).select('dioId', 'roomId', 'name');
}

export async function findById(dioId: number): Promise<Dio> {
  return knex(TABLE).first('dioId', 'roomId', 'name').where('dioId', dioId);
}

export async function findByRoomId(roomId: string): Promise<Dio[]> {
  return knex(TABLE).select('dioId', 'roomId', 'name').where('roomId', roomId);
}

export async function save(newDio: Dio): Promise<Dio> {
  return knex(TABLE)
    .insert(newDio)
    .then(() => newDio);
}

export function remove(dioId: number) {
  return knex(TABLE).where('dioId', dioId).del();
}

export function findByIdAndUpdate(
  dioId: number,
  payload: Partial<Omit<Dio, 'dioId'>>,
) {
  return knex(TABLE).update(payload).where('dioId', dioId);
}
