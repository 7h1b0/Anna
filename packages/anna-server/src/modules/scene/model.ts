import Ajv from 'ajv';
import { v4 as uuidv4 } from 'uuid';
import knex from '../../knexClient';
import sceneSchema from './schema';
import { TABLE as ACTION_TABLE, findBySceneId } from './action';
import { ToggleDio, ToggleHueLight } from 'utils/actions';

export const TABLE = 'scenes';
export const COLUMNS = [
  'sceneId',
  'name',
  'description',
  'favorite',
  'createdAt',
  'updatedAt',
  'createdBy',
];

export type Scene = {
  sceneId: string;
  name: string;
  description: string;
  createdBy: string;
  favorite: boolean;
  createdAt?: number;
  updatedAt?: number;
};

export type SceneAction = {
  actions: (ToggleDio | ToggleHueLight)[];
} & Scene;

export function validate(data: Record<string, unknown>) {
  const ajv = new Ajv();
  return ajv.validate(sceneSchema, data);
}

export async function findAll(): Promise<Scene[]> {
  return knex(TABLE).select(COLUMNS);
}

export async function findById(
  sceneId: string,
): Promise<SceneAction | undefined> {
  const fetchScene = knex(TABLE).first(COLUMNS).where('sceneId', sceneId);
  const fetchActions = findBySceneId(sceneId);

  const [scene, actions] = await Promise.all([fetchScene, fetchActions]);
  if (scene) {
    return { ...scene, actions };
  }
  return;
}

export async function findByIdAndUpdate({
  sceneId,
  name,
  description,
  favorite,
  actions,
}: SceneAction) {
  return knex
    .transaction(async (trx) => {
      return trx
        .update({ name, description, favorite })
        .where('sceneId', sceneId)
        .into(TABLE)
        .then((row) => {
          if (row < 1) {
            throw new Error('No scene found');
          }
          return trx.del().where('sceneId', sceneId).into(ACTION_TABLE);
        })
        .then(() => {
          return Promise.all(
            actions.map((action) => {
              const actionId = uuidv4();
              const formatedAction = {
                type: action.type,
                targetId: action.targetId,
                sceneId,
                actionId,
                body: JSON.stringify(action.body),
              };
              return trx.insert(formatedAction).into(ACTION_TABLE);
            }),
          );
        });
    })
    .catch((err: Error) => {
      if (err.message === 'No scene found') {
        return 0;
      }
      throw err;
    });
}

export async function save({
  name,
  description,
  favorite,
  createdBy,
  actions,
}: Omit<SceneAction, 'sceneId'>) {
  const sceneId = uuidv4();
  await knex.transaction((trx) => {
    const insertIntoScene = trx
      .insert({
        sceneId,
        description,
        favorite,
        name,
        createdBy,
      })
      .into(TABLE);

    const insertIntoActions = actions.map((action) => {
      const formatedAction = {
        type: action.type,
        targetId: action.targetId,
        sceneId,
        body: JSON.stringify(action.body),
        actionId: uuidv4(),
      };
      return trx.insert(formatedAction).into(ACTION_TABLE);
    });

    return Promise.all([insertIntoScene, ...insertIntoActions]);
  });

  return sceneId;
}

export async function remove(sceneId: string): Promise<number[]> {
  return knex.transaction((trx) => {
    const removeScene = trx.del().where('sceneId', sceneId).into(TABLE);

    const removeActions = trx
      .del()
      .where('sceneId', sceneId)
      .into(ACTION_TABLE);

    return Promise.all([removeScene, removeActions]);
  });
}

export async function findAllFavorite(): Promise<Scene[]> {
  return knex(TABLE).select(COLUMNS).where('favorite', true);
}
