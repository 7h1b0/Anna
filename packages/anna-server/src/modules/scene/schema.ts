import TYPES from 'utils/type';

export default {
  $id: 'scene',
  type: 'object',
  additionalProperties: false,
  properties: {
    sceneId: {
      type: 'string',
      pattern: '[a-z0-9-]{36}',
    },
    name: {
      type: 'string',
      minLength: 3,
    },
    description: {
      type: 'string',
      minLength: 5,
    },
    favorite: {
      type: 'boolean',
    },
    actions: {
      type: 'array',
      minItems: 1,
      items: [
        {
          type: 'object',
          additionalProperties: false,
          properties: {
            targetId: {
              type: 'number',
            },
            actionId: {
              type: 'string',
              pattern: '[a-z0-9-]{36}',
            },
            name: {
              type: 'string',
              minLength: 3,
            },
            type: {
              type: 'string',
              enum: [TYPES.DIO, TYPES.HUE_LIGHT],
            },
            body: {
              type: 'object',
              minProperties: 1,
            },
          },
          required: ['targetId', 'body', 'type'],
        },
      ],
    },
  },
  required: ['name', 'actions', 'favorite'],
};
