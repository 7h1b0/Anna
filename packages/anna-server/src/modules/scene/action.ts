import knex from '../../knexClient';
import TYPES from 'utils/type';
import { ToggleDio, ToggleHueLight, AnnaAction } from 'utils/actions';
import * as logger from 'utils/logger';

export const TABLE = 'actions';

export type HueLigthBody = {
  on: boolean;
  bri?: number;
  xy?: number[];
};

export type DioBody = { on: boolean };

export type Action = {
  sceneId: string;
  name: string;
  targetId: string;
  readonly type: TYPES.DIO;
  body: string;
};

export async function findBySceneId(sceneId: string): Promise<AnnaAction[]> {
  return knex(TABLE)
    .select(['type', 'body', 'targetId'])
    .where('sceneId', sceneId)
    .then((entries: Action[]) => {
      return entries.reduce((acc: AnnaAction[], entry: Action) => {
        try {
          const body = JSON.parse(entry.body);
          if (entry.type === TYPES.DIO) {
            return acc.concat([new ToggleDio(entry.targetId, body)]);
          }
          return acc.concat(new ToggleHueLight(entry.targetId, body));
        } catch (e) {
          logger.error(`Body of entry #${entry.name} is malformed`);
          return acc;
        }
      }, []);
    });
}
