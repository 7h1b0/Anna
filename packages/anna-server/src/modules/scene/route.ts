import { Router } from 'express';
import * as Scene from '../../modules/scene/model';
import { callScene } from '../../utils/actions';
import dispatch from '../../utils/dispatch';
import * as logger from '../../utils/logger';

const routes = Router();
routes
  .route('/api/scenes')
  .get((req, res) => {
    Scene.findAll()
      .then((scenes) => res.json(scenes))
      .catch((err) => res.status(500).send({ err }));
  })
  .post((req, res) => {
    const isValid = Scene.validate(req.body);
    if (!isValid) {
      res.sendStatus(400);
    } else {
      const createdBy = res.locals.user.userId;
      Scene.save({ ...req.body, createdBy })
        .then((sceneId) => res.status(201).json({ sceneId }))
        .catch((err) => res.status(500).send({ err }));
    }
  });

routes
  .route('/api/scenes/:sceneId([a-fA-F0-9-]{36})')
  .get((req, res) => {
    Scene.findById(req.params.sceneId)
      .then((scene) => {
        if (!scene) {
          res.sendStatus(404);
        } else {
          res.json(scene);
        }
      })
      .catch((err) => {
        res.status(500).send({ err });
      });
  })
  .patch((req, res) => {
    const isValid = Scene.validate(req.body);
    if (!isValid) {
      res.sendStatus(400);
    } else {
      Scene.findByIdAndUpdate({ ...req.body, sceneId: req.params.sceneId })
        .then((rowsAffected) => {
          if (rowsAffected) {
            res.sendStatus(204);
          } else {
            res.sendStatus(404);
          }
        })
        .catch((err) => res.status(500).send({ err }));
    }
  })
  .delete((req, res) => {
    Scene.remove(req.params.sceneId)
      .then(([removedScene]) => {
        if (removedScene < 1) {
          res.sendStatus(404);
        } else {
          res.sendStatus(204);
        }
      })
      .catch((err) => {
        res.status(500).send({ err });
      });
  });

routes.get('/api/scenes/:sceneId([a-fA-F0-9-]{36})/action', (req, res) => {
  dispatch(callScene(req.params.sceneId))
    .then(() => res.end())
    .catch((err) => {
      logger.error(err);
      res.status(500).send({ err });
    });
});

routes.get('/api/scenes/favorites', (req, res) => {
  Scene.findAllFavorite()
    .then((scenes) => res.json(scenes))
    .catch((err) => res.status(500).send({ err }));
});

routes.get(
  '/api/scenes/:sceneId([a-fA-F0-9-]{36})/favorite/:state(1|0)',
  (req, res) => {
    Scene.findByIdAndUpdateFavoriteState(
      req.params.sceneId,
      req.params.state === '1',
    )
      .then((rowAffected) => {
        if (rowAffected) {
          res.sendStatus(204);
        } else {
          res.sendStatus(404);
        }
      })
      .catch((err) => res.status(500).send({ err }));
  },
);

export default routes;
