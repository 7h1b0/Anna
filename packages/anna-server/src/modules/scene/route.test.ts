import request from 'supertest';
import { createUser } from 'factories';
import knex from '../../knexClient';
import * as Scene from '../../modules/scene/model';
import * as Action from '../../modules/scene/action';
import * as User from '../../modules/user/model';
import app from '../../index';
import dispatch from '../../utils/dispatch';

jest.mock('../../utils/dispatch');

const user = createUser({ userId: '29699398-449c-48fb-8f5c-84186cdf8279' });
const initScenes = [
  {
    sceneId: '05442486-0878-440c-9db1-a7006c25a39f',
    description: 'this is a test',
    name: 'scene_1',
    favorite: true,
    createdBy: '29699398-449c-48fb-8f5c-84186cdf8279',
    createdAt: new Date('2018-01-01'),
    updatedAt: new Date('2018-01-02'),
  },
  {
    sceneId: '18feb598-32cb-472f-8b29-a7e7fe41e06b',
    description: 'this is a second test',
    name: 'scene_2',
    favorite: false,
    createdBy: '29699398-449c-48fb-8f5c-84186cdf8279',
    createdAt: new Date('2018-01-01'),
    updatedAt: new Date('2018-01-02'),
  },
];

const initActions = [
  {
    actionId: '0fc1d78e-fd1c-4717-b610-65d2fa3d01b2',
    sceneId: '05442486-0878-440c-9db1-a7006c25a39f',
    type: 'DIO',
    name: 'action turn on',
    targetId: 1,
    body: JSON.stringify({ on: true }),
  },
  {
    actionId: '1fc1d78e-fd1c-4717-b610-65d2fa3d01b2',
    sceneId: '05442486-0878-440c-9db1-a7006c25a39f',
    type: 'DIO',
    name: 'action turn off',
    targetId: 2,
    body: JSON.stringify({ on: false }),
  },
  {
    actionId: '2fc1d78e-fd1c-4717-b610-65d2fa3d01b2',
    sceneId: '18feb598-32cb-472f-8b29-a7e7fe41e06b',
    type: 'HUE_LIGHT',
    name: 'call scene',
    targetId: 1,
    body: null,
  },
];

describe('Scene API', () => {
  beforeAll(async () => {
    await knex(User.TABLE).truncate();
    await knex(User.TABLE).insert(user);
  });

  beforeEach(async () => {
    await knex(Action.TABLE).truncate();
    await knex(Action.TABLE).insert(initActions);

    await knex(Scene.TABLE).truncate();
    await knex(Scene.TABLE).insert(initScenes);
  });

  describe('/api/scenes', () => {
    describe('GET', () => {
      it('should return all scene', async () => {
        const response = await request(app)
          .get('/api/scenes')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.body).toMatchSnapshot();
      });

      it('should retun 401 when user is not authenticated', async () => {
        const response = await request(app)
          .get('/api/scenes')
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBeUnauthorized();
      });
    });

    describe('POST', () => {
      it('should return 400 if scene is malformed', async () => {
        const scene = {
          description: 'this is a test scene',
          actions: [
            {
              targetId: 2,
              name: 'action',
              type: 'DIO',
              body: {
                on: true,
              },
            },
          ],
        };

        const response = await request(app)
          .post('/api/scenes')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send(scene);

        expect(response.status).toBeBadRequest();
      });

      it('should create a scene', async () => {
        const scene = {
          name: 'testScene',
          description: 'this is a test scene',
          favorite: true,
          actions: [
            {
              targetId: 1,
              name: 'action',
              type: 'DIO',
              body: {
                on: true,
              },
            },
          ],
        };

        const response = await request(app)
          .post('/api/scenes')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send(scene);

        expect(response.status).toHaveStatusOk();

        const sceneFromDatabase = await knex(Scene.TABLE)
          .first()
          .where('sceneId', response.body.sceneId);

        expect(sceneFromDatabase).toMatchSnapshot({
          sceneId: expect.stringMatching(/[a-fA-F0-9-]{36}/),
          createdAt: expect.any(Number),
          updatedAt: expect.any(Number),
        });

        const actions = await knex(Action.TABLE)
          .select()
          .where('sceneId', response.body.sceneId);

        expect(actions).toHaveLength(1);
        expect(actions[0]).toMatchSnapshot({
          actionId: expect.stringMatching(/[a-fA-F0-9-]{36}/),
          sceneId: expect.stringMatching(/[a-fA-F0-9-]{36}/),
        });
      });

      it('should retun 401 when user is not authenticated', async () => {
        const response = await request(app)
          .post('/api/scenes')
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBeUnauthorized();
      });
    });
  });

  describe('/api/scenes/:id', () => {
    describe('GET', () => {
      it('should retun a scene', async () => {
        const response = await request(app)
          .get(`/api/scenes/${initScenes[0].sceneId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.status).toHaveStatusOk();
        expect(response.body).toMatchSnapshot();
      });

      it('should retun 401 when user is not authenticated', async () => {
        const response = await request(app)
          .get(`/api/scenes/${initScenes[1].sceneId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBeUnauthorized();
      });

      it("should retun 404 when a scene don't exist", async () => {
        const response = await request(app)
          .get('/api/scenes/3fc1d78e-fd1c-4717-b610-65d2fa3d01b2')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.status).toBe(404);
      });
    });

    describe('DELETE', () => {
      it('should delete a scene and all associated actions', async () => {
        const response = await request(app)
          .delete(`/api/scenes/${initScenes[0].sceneId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.status).toHaveStatusOk();

        const scene = await knex(Scene.TABLE)
          .select()
          .where('sceneId', initScenes[0].sceneId);

        expect(scene).toHaveLength(0);

        const actions = await knex(Action.TABLE)
          .select()
          .where('sceneId', initScenes[0].sceneId);

        expect(actions).toHaveLength(0);
      });

      it('should return 404 if scene is not found', async () => {
        const response = await request(app)
          .delete('/api/scenes/5fc1d78e-fd1c-4717-b610-65d2fa3d01b2')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.status).toBe(404);
      });

      it('should retun 401 when user is not authenticated', async () => {
        const response = await request(app)
          .delete(`/api/scenes/${initScenes[1].sceneId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBeUnauthorized();
      });
    });

    describe('PATCH', () => {
      it('should return 400 if scene is malformed', async () => {
        const scene = {
          description: 'this is an updated scene',
          actions: [
            {
              targetId: 2,
              name: 'action',
              type: 'DIO',
              body: {
                on: true,
              },
            },
          ],
        };

        const response = await request(app)
          .patch(`/api/scenes/${initScenes[1].sceneId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send(scene);

        expect(response.status).toBeBadRequest();
      });

      it('should retun 401 when user is not authenticated', async () => {
        const response = await request(app)
          .patch(`/api/scenes/${initScenes[1].sceneId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBeUnauthorized();
      });

      it('should return 404 if scene is not found', async () => {
        const scene = {
          description: 'this is an updated scene',
          name: 'testtest',
          favorite: true,
          actions: [
            {
              targetId: 2,
              name: 'action',
              type: 'DIO',
              body: {
                on: true,
              },
            },
          ],
        };
        const response = await request(app)
          .patch('/api/scenes/faaed78e-fd1c-4717-b610-65d2fa3d01b2')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send(scene);

        expect(response.status).toBe(404);
      });

      it('should update a scene', async () => {
        const updatedScene = {
          description: 'this is an updated test',
          name: 'scene_2',
          favorite: false,
          actions: [
            {
              type: 'HUE_LIGHT',
              name: 'call scene',
              targetId: 3,
              body: { on: true },
            },
          ],
        };

        const response = await request(app)
          .patch(`/api/scenes/${initScenes[1].sceneId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send(updatedScene);

        expect(response.status).toHaveStatusOk();

        const sceneFromDatabase = await knex(Scene.TABLE)
          .first()
          .where('sceneId', initScenes[1].sceneId);

        expect(sceneFromDatabase).toMatchSnapshot({
          sceneId: expect.stringMatching(/[a-fA-F0-9-]{36}/),
          createdAt: expect.any(Number),
          updatedAt: expect.any(Number),
        });
      });
    });
  });

  describe('/api/scenes/:id_scene/action', () => {
    it('should call dispatch', async () => {
      const response = await request(app)
        .get(`/api/scenes/${initScenes[1].sceneId}/action`)
        .set('Accept', 'application/json')
        .set('x-access-token', user.token);

      expect(response.status).toHaveStatusOk();
      expect(dispatch).toHaveBeenCalledWith({
        type: 'SCENE',
        targetId: initScenes[1].sceneId,
      });
    });

    it('should retun 401 when user is not authenticated', async () => {
      const response = await request(app)
        .get(`/api/scenes/${initScenes[0].sceneId}/action`)
        .set('Accept', 'application/json')
        .set('x-access-token', 'fake');

      expect(response.status).toBeUnauthorized();
    });
  });

  describe('/api/scenes/favorites', () => {
    it('should return all favorite scenes', async () => {
      const response = await request(app)
        .get(`/api/scenes/favorites`)
        .set('Accept', 'application/json')
        .set('x-access-token', user.token);

      expect(response.status).toHaveStatusOk();
      expect(response.body).toHaveLength(1);
    });

    it('should retun 401 when user is not authenticated', async () => {
      const response = await request(app)
        .get(`/api/scenes/favorites`)
        .set('Accept', 'application/json')
        .set('x-access-token', 'fake');

      expect(response.status).toBeUnauthorized();
    });
  });

  describe('/api/scenes/:sceneId/favorite/:state', () => {
    it('should update favorite state', async () => {
      const scene = initScenes[1];
      const response = await request(app)
        .get(`/api/scenes/${scene.sceneId}/favorite/1`)
        .set('Accept', 'application/json')
        .set('x-access-token', user.token);

      expect(response.status).toHaveStatusOk();

      const updatedScene = await knex(Scene.TABLE)
        .first('favorite')
        .where('sceneId', scene.sceneId);
      expect(updatedScene.favorite).toBe(true);
    });

    it('should retun 401 when user is not authenticated', async () => {
      const response = await request(app)
        .get(`/api/scenes/${initScenes[1].sceneId}/favorite/1`)
        .set('Accept', 'application/json')
        .set('x-access-token', 'fake');

      expect(response.status).toBeUnauthorized();
    });

    it('should retun 404 when scene is not found', async () => {
      const response = await request(app)
        .get(`/api/scenes/5fc1d78e-fd1c-4717-b610-65d2fa3d01b2/favorite/1`)
        .set('Accept', 'application/json')
        .set('x-access-token', 'fake');

      expect(response.status).toBeUnauthorized();
    });
  });
});
