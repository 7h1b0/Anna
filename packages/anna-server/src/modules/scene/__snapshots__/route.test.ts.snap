// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scene API /api/scenes GET should return all scene 1`] = `
Array [
  Object {
    "createdAt": 1514764800000,
    "createdBy": "29699398-449c-48fb-8f5c-84186cdf8279",
    "description": "this is a test",
    "favorite": true,
    "name": "scene_1",
    "sceneId": "05442486-0878-440c-9db1-a7006c25a39f",
    "updatedAt": 1514851200000,
  },
  Object {
    "createdAt": 1514764800000,
    "createdBy": "29699398-449c-48fb-8f5c-84186cdf8279",
    "description": "this is a second test",
    "favorite": false,
    "name": "scene_2",
    "sceneId": "18feb598-32cb-472f-8b29-a7e7fe41e06b",
    "updatedAt": 1514851200000,
  },
]
`;

exports[`Scene API /api/scenes POST should create a scene 1`] = `
Object {
  "createdAt": Any<Number>,
  "createdBy": "29699398-449c-48fb-8f5c-84186cdf8279",
  "description": "this is a test scene",
  "favorite": true,
  "name": "testScene",
  "sceneId": StringMatching /\\[a-fA-F0-9-\\]\\{36\\}/,
  "updatedAt": Any<Number>,
}
`;

exports[`Scene API /api/scenes POST should create a scene 2`] = `
Object {
  "actionId": StringMatching /\\[a-fA-F0-9-\\]\\{36\\}/,
  "body": "{\\"on\\":true}",
  "name": null,
  "sceneId": StringMatching /\\[a-fA-F0-9-\\]\\{36\\}/,
  "targetId": 1,
  "type": "DIO",
}
`;

exports[`Scene API /api/scenes/:id GET should retun a scene 1`] = `
Object {
  "actions": Array [
    Object {
      "body": Object {
        "on": true,
      },
      "targetId": 1,
      "type": "DIO",
    },
    Object {
      "body": Object {
        "on": false,
      },
      "targetId": 2,
      "type": "DIO",
    },
  ],
  "createdAt": 1514764800000,
  "createdBy": "29699398-449c-48fb-8f5c-84186cdf8279",
  "description": "this is a test",
  "favorite": true,
  "name": "scene_1",
  "sceneId": "05442486-0878-440c-9db1-a7006c25a39f",
  "updatedAt": 1514851200000,
}
`;

exports[`Scene API /api/scenes/:id PATCH should update a scene 1`] = `
Object {
  "createdAt": Any<Number>,
  "createdBy": "29699398-449c-48fb-8f5c-84186cdf8279",
  "description": "this is an updated test",
  "favorite": false,
  "name": "scene_2",
  "sceneId": StringMatching /\\[a-fA-F0-9-\\]\\{36\\}/,
  "updatedAt": Any<Number>,
}
`;
